Source/WebCore/dom/TrustedTypePolicyFactory.cpp

```diff
@@ -100,20 +104,126 @@ Ref<TrustedScript> TrustedTypePolicyFactory::emptyScript() const
+
+    AtomString elementNs = elementNamespace.isEmpty() ? HTMLNames::xhtmlNamespaceURI : AtomString(elementNamespace);
+    AtomString attrNs = attrNamespace.isEmpty() ? nullAtom() : AtomString(attrNamespace);
+
+    Vector<AttributeTypeListItem> table = attributeTypeVector();
+
+    for (const auto& entry : table) {
+        bool entryMatches = (entry.element == AtomString(localName)
+            && entry.elementNs == elementNs
+            && entry.attribute == attribute
+            && entry.attributeNs == attrNs);
+        if (entryMatches)
+            return String(entry.type);
+    }
+
     return String();
 }
 
-String TrustedTypePolicyFactory::getPropertyType(const String& tagName, const String& attribute, const String& elementNamespace) const
+struct PropertyTypeListItem {
+    AtomString element;
+    AtomString property;
+    AtomString elementNs;
+    AtomString type;
+};
+
+static const Vector<PropertyTypeListItem> propertyTypeVector()
+{
+    static MainThreadNeverDestroyed<Vector<PropertyTypeListItem>> result;
+    const QualifiedName anyElement(nullAtom(), AtomString("*"_s), nullAtom());
+
+    const struct {
+        const QualifiedName element;
+        const String property;
+        const String type;
+    } typeTable[] = {
+        { HTMLNames::embedTag, "src"_s, "TrustedScriptURL"_s },
+        { HTMLNames::iframeTag, "srcdoc"_s, "TrustedHTML"_s },
+        { HTMLNames::objectTag, "codeBase"_s, "TrustedScriptURL"_s },
+        { HTMLNames::objectTag, "data"_s, "TrustedScriptURL"_s },
+        { HTMLNames::scriptTag, "innerText"_s, "TrustedScript"_s },
+        { HTMLNames::scriptTag, "src"_s, "TrustedScriptURL"_s },
+        { HTMLNames::scriptTag, "text"_s, "TrustedScript"_s },
+        { HTMLNames::scriptTag, "textContent"_s, "TrustedScript"_s },
+        { anyElement, "innerHTML"_s, "TrustedHTML"_s },
+        { anyElement, "outerHTML"_s, "TrustedHTML"_s },
+    };
+
+    for (const auto& entry : typeTable) {
+        result->append(PropertyTypeListItem {
+            entry.element.localName(), AtomString(entry.property), entry.element.namespaceURI(), AtomString(entry.type)
+        });
+    }
+
+    return result;
+}
+
+String TrustedTypePolicyFactory::getPropertyType(const String& tagName, const String& property, const String& elementNamespace) const
 {
-    UNUSED_PARAM(tagName);
-    UNUSED_PARAM(attribute);
-    UNUSED_PARAM(elementNamespace);
+    auto localName = tagName.convertToASCIILowercase();
+    AtomString elementNs = elementNamespace.isEmpty() ? HTMLNames::xhtmlNamespaceURI : AtomString(elementNamespace);
+
+    if (property.startsWith("on"_s))
+        return "TrustedScript"_s;
+
+    Vector<PropertyTypeListItem> table = propertyTypeVector();
+
+    for (const auto& entry : table) {
+        bool elementMatches = entry.element == AtomString(localName) && entry.elementNs == elementNs;
+        bool entryMatches = (elementMatches
+            || entry.element == AtomString("*"_s))
```

Comment:
Let's not convert "*" to AtomString in every iteration. We should just use null string instead.